text/0000-discriminant-intrinsic.md

```diff
@@ -0,0 +1,339 @@
+types.
+
+# Motivation
+
+Many operations that work with discriminant values can be significantly improved with the ability to
+extract the value of the discriminant that is used to distinguish between variants in an enum. While
+trivial cases often optimise well, more complex ones would benefit from direct access to this value.
+
+A good example is the `SqlState` enum from the `postgres` crate (Listed at the end of this RFC). It
+contains 233 variants, of which all but one contain no fields. The most obvious implementation of
+(for example) the `PartialEq` trait looks like this:
+
+```rust
+match (self, other) {
+    (&Unknown(ref s1), &Unknown(ref s2)) => s1 == s2,
+    (&SuccessfulCompletion, &SuccessfulCompletion) => true,
+    (&Warning, &Warning) => true,
+    (&DynamicResultSetsReturned, &DynamicResultSetsReturned) => true,
+    (&ImplicitZeroBitPadding, &ImplicitZeroBitPadding) => true,
+	      .
+		  .
+		  .
+    (_, _) => false
+}
+```
+
+Even with optimisations enabled, this code is very suboptimal, producing
+[this code](https://gist.github.com/Aatch/c23a45634b10aaecad05). A way to extract the discriminant
+would allow this code:
+
+```rust
+match (self, other) {
+    (&Unknown(ref s1), &Unknown(ref s2)) => s1 == s2,
+    (l, r) => unsafe {
+	    discriminant_value(l) == discriminant(r)
+	}
+}
+```
+
+Which is compiled into [this IR](https://gist.github.com/Aatch/beb736b93a908aa67e84).
+
+# Detailed design
+
+## What is a discriminant?
+
+A discriminant is a value stored in an enum type that indicates which variant the value is. The most
+common case is that the discriminant is stored directly as an extra field in the variant. However,
+the discriminant may be stored in any place, and in any format. However, we can always extract the
+discriminant from the value somehow.
+
+## Implementation
+
+For any given type, `discriminant_value` will return a `u64` value. The values returned are as
+specified:
+
+* **Non-Enum Type**: Always 0
+* **Enum Type**: A value that uniquely identifies that variant within its type. I.E. for a given
+  enum typem `E`, and two values of type `E`, `a` and `b`, the expression `discriminant_value(a) ==
+  discriminant_value(b)` is true iff `a` and `b` are the same variant. Two values of different types
+  may return the same discriminant value.
+
+The reason for this specification is to allow flexibilty in usage of the intrinsic without
+compromising our ability to change the representation at will.
+
+# Drawbacks
+
+* Potentially exposes implementation details. However, relying the specific values returned from
+`discriminant_value` should be considered bad practice, as the intrinsic provides no such guarantee.
+
+* Does not allow for the value to be used as part of ordering.
+
+* Allows non-enum types to be provided. This may be unexpected by some users.
+
+# Alternatives
+
+* More strongly specify the values returned. This would allow for a broader range of uses, but
```

Comment:
As the RFC is currently written, the only guarantee you have is that two values of the same type produce the same value, iff the variants are the same. While you could implement *a* ordering with this, I deliberately restricted such that the ordering would be effectively unspecified. The restriction could be lifted, but it would require a more rigorous specification of discriminant values.